File: 236-lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':


        def LCA(root,p,q):
            if root==None:
                return root
            
            if root==p:
                return root
            elif root==q:
                return root

            left_side=LCA(root.left,p,q)
            right_side=LCA(root.right,p,q)

            if left_side!=None and right_side!=None:
                return root
            
            return left_side if right_side==None else right_side
        return LCA(root,p,q)


        
        
            
            





            
        

File: 236-lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode searchq (TreeNode root , TreeNode p ,TreeNode q )
    {
        if(root==null)
        {
            return null;
        }
        if(root==p)
        {
            return root;
        }
        if(root == q)
        {
            return root;
        }
        TreeNode left =searchq(root.left ,p,q);
        TreeNode right=searchq(root.right,p,q);
        if(left!=null && right!=null)
        {
            return root;
        }
        return left==null?right:left;
    }
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
     return searchq(root,p,q);
    }
}